server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/AbstractRecipientRewriteTableMailet.java

```diff
@@ -123,75 +127,158 @@ public void service(Mail mail) throws MessagingException {
+    private Function<Mapping, Mapping> appendDefaultDomain(final String defaultDomain) {
+        return new Function<Mapping, Mapping>() {
+            @Override
+            public Mapping apply(Mapping input) {
+                return input.appendDomain(defaultDomain);
             }
-        }
+        };
+    }
+
+    private Predicate<MailAddress> isLocalServer(final MailetContext context) {
+        return new Predicate<MailAddress>() {
+            @Override
+            public boolean apply(MailAddress input) {
+                return context.isLocalServer(input.getDomain());
+            }
+        };
+    }
 
-        if (remoteRecipients.size() > 0) {
+    private Predicate<MailAddress> isNotLocalServer(final MailetContext context) {
+        return new Predicate<MailAddress>() {
+            @Override
+            public boolean apply(MailAddress input) {
+                return !context.isLocalServer(input.getDomain());
+            }
+        };
+    }
+
+    private void forwardToRemoteAddress(MailAddress sender, MailAddress recipient, MimeMessage message, ImmutableList<MailAddress> mailAddresses, final MailetContext context) throws MessagingException {
+        ImmutableList<MailAddress> remoteAddress = FluentIterable.from(mailAddresses)
+            .filter(isNotLocalServer(context))
+            .toList();
+
+        if (!remoteAddress.isEmpty()) {
+            StringBuilder logBuffer = new StringBuilder(128);
             try {
-                getMailetContext().sendMail(sender, remoteRecipients, message);
-                StringBuilder logBuffer = new StringBuilder(128).append("Mail for ").append(recipient).append(" forwarded to ");
-                for (Iterator<MailAddress> j = remoteRecipients.iterator(); j.hasNext();) {
-                    logBuffer.append(j.next());
-                    if (j.hasNext())
-                        logBuffer.append(", ");
-                }
-                getMailetContext().log(LogLevel.INFO, logBuffer.toString());
+                context.sendMail(sender, remoteAddress, message);
+                logBuffer.append("Mail for ")
+                    .append(recipient).append(" forwarded to ")
+                    .append(remoteAddress);
             } catch (MessagingException me) {
-                StringBuilder logBuffer = new StringBuilder(128).append("Error forwarding mail to ");
-                for (Iterator<MailAddress> j = remoteRecipients.iterator(); j.hasNext();) {
-                    logBuffer.append(j.next());
-                    if (j.hasNext())
-                        logBuffer.append(", ");
-                }
-                logBuffer.append("attempting local delivery");
-
-                getMailetContext().log(LogLevel.INFO, logBuffer.toString());
+                logBuffer.append("Error forwarding mail to ")
+                    .append(remoteAddress)
+                    .append("attempting local delivery");
                 throw me;

```

Comment:
You should not throw here ! If you throw here, the following scenario can happen : 
 - A send mail to B and C (A, B, C are locals)
 - B gets re-written into D (D is not local)
 - James fails while sending mail to D (for any reason)
 - C do not get its e-mail (but it could have !)

You also need a test about that